stable_baselines3/dqn/dqn.py

```diff
@@ -208,8 +209,20 @@ def train(self, gradient_steps: int, batch_size: int = 100) -> None:
             # Retrieve the q-values for the actions from the replay buffer
             current_q_values = th.gather(current_q_values, dim=1, index=replay_data.actions.long())
 
-            # Compute Huber loss (less sensitive to outliers)
-            loss = F.smooth_l1_loss(current_q_values, target_q_values)
+            # Special case when using PrioritizedReplayBuffer (PER)
+            if isinstance(self.replay_buffer, PrioritizedReplayBuffer):
+                # TD error in absolute value
+                td_error = th.abs(current_q_values - target_q_values)
+                # Weighted Huber loss using importance sampling weights
+                loss = (replay_data.weights * th.where(td_error < 1.0, 0.5 * td_error**2, td_error - 0.5)).mean()
+                # Update priorities, they will be proportional to the td error
+                assert replay_data.leaf_nodes_indices is not None, "Node leaf node indices provided"
+                self.replay_buffer.update_priorities(
+                    replay_data.leaf_nodes_indices, td_error, self._current_progress_remaining
+                )
+            else:
+                # Compute Huber loss (less sensitive to outliers)
+                loss = F.smooth_l1_loss(current_q_values, target_q_values)
```

Comment:
@AlexPasqua Ideally, we'd like to be able to associate it with all off-policy algo's without adaptation, but I don't see a simple way of doing it at this stage.
Also related, we had discussed not modifying DQN: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/127#issuecomment-1381530029

Reply:
I'm interested in this PR. Since every algo-specific `train` method includes a `replay_buffer.sample` line, couldn't we just additionally add a `replay_buffer.update` line? The update function could take in the current and target q values whenever a value function is present or maybe even all the local variables. It would do nothing for the vanilla replay buffer. Would this be an acceptable modification?

Reply:
Thanks for your comment!
How do you handle the loss in your proposal?

Reply:
If we want this to work for general off-policy algorithms, we could update the ReplayBufferSample-like classes to additionally include an `importance_sampling_weight` attribute which would be updated from the `replay_buffer.update` method.

Then I see two ways to handle the loss under this interface:

1. Estimate TD error from the loss as such:
```
losses = loss_fn(current_q_values, target_q_values, reduction='none')

# e.g. If loss is L2, then it's basically th.sqrt(loss). If loss is L1, td_error = loss
td_error = importance_sampling_weight * function_to_approx_td_error(losses)  

loss = losses.mean()
```
Obviously the downside of this is that it requires hand engineering for the different types of loss functions or priority metrics.

2. Make any value-based `train` methods "td-error" centric in the sense that we always compute `td_error = importance_sampling_weight * th.abs(current_q_values - target_q_values)` first, then the loss `loss = loss_fn(td_error)`. The downsides of this approach is that we cant use the pytorch api for computing the loss, and would have to write functions for those.

Either approach requires computing a `td_error` variable which unfortunately requires somewhat intrusive code changes. What do you think?

Reply:
maybe to make things clearer: my plan is not to have for all algorithms, mainly for two reasons:

1. Keep the code concise (in fact, I would like to have RAINBOW and keep vanilla DQN, see https://github.com/DLR-RM/stable-baselines3/issues/622)
2. I don't think it works for entropy-RL algorithms (SAC and derivates), so it would be limited to DQN/QR-DQN and TD3

Reply:
In that case, I'd be happy to contribute to the Rainbow DQN feature request. Dueling DQN and double DQN are straightforward to implement, and @araffin PER implementation seems to work. Also C51 is already implemented in contrib. We would just need n-step returns and noisy nets right?

Reply:
"just" yes, I would be happy to receive such PR =)
the main thing is to benchmark the implementation and reproduce the published results.
This PR is also still open because I was not satisfied by the result of DQN + PER (I couldn't see significant different with respect to DQN).

Reply:
One thing I had in mind was to implement CNN for SBX (https://github.com/araffin/sbx) in order to iterate faster and check the PER, but I had no time to do so until now...

Reply:
Why don't we implement the toy environment from figure 1 of https://arxiv.org/pdf/1511.05952 as the PER benchmark? It would be a simpler initial check for correctness than the Atari environments

Reply:
The toy environment can be a start for fast iteration and debugging, but what we learn in the past is that subtle bugs only show up when doing more complex task (see https://github.com/DLR-RM/stable-baselines3/issues/48 and https://github.com/DLR-RM/stable-baselines3/issues/47 where we found bugs like PyTorch and TF RMSProp are not the same)

Reply:
I see, will definitely work towards it!